esphome/components/shelly_dimmer/light.py

```diff
@@ -43,12 +46,108 @@
 CONF_NRST_PIN = "nrst_pin"
 CONF_BOOT0_PIN = "boot0_pin"
 
+KNOWN_FIRMWARE = {
+    "51.5": (
+        "https://github.com/jamesturton/shelly-dimmer-stm32/releases/download/v51.5/shelly-dimmer-stm32_v51.5.bin",
+        "553fc1d78ed113227af7683eaa9c26189a961c4ea9a48000fb5aa8f8ac5d7b60",
+    ),
+    "51.6": (
+        "https://github.com/jamesturton/shelly-dimmer-stm32/releases/download/v51.6/shelly-dimmer-stm32_v51.6.bin",
+        "eda483e111c914723a33f5088f1397d5c0b19333db4a88dc965636b976c16c36",
+    ),
+}
+
+
+def parse_firmware_version(value):
+    match = re.match(r"(\d+).(\d+)", value)
+    if match is None:
+        raise ValueError(f"Not a valid version number {value}")
+    major = int(match[1])
+    minor = int(match[2])
+    return major, minor
+
+
+def get_firmware(value):
+    if not value[CONF_UPDATE]:
+        return None
+
+    def dl(url):
+        try:
+            req = requests.get(url)
+            req.raise_for_status()
+        except requests.exceptions.RequestException as e:
+            raise cv.Invalid(f"Could not download firmware file ({url}): {e}")
+
+        h = hashlib.new("sha256")
+        h.update(req.content)
+        return req.content, h.hexdigest()
+
+    url = value[CONF_URL]
+
+    if CONF_SHA256 in value:  # we have a hash, enable caching
+        path = (
+            Path(CORE.config_dir)
+            / ".esphome"
+            / DOMAIN
+            / (value[CONF_SHA256] + "_fw_stm.bin")
+        )
+
+        if not path.is_file():
+            firmware_data, dl_hash = dl(url)
+
+            if dl_hash != value[CONF_SHA256]:
+                raise cv.Invalid(
+                    f"Hash mismatch for {url}: {dl_hash} != {value[CONF_SHA256]}"
+                )
+
+            path.parent.mkdir(exist_ok=True, parents=True)
+            path.write_bytes(firmware_data)
+
+        else:
+            firmware_data = path.read_bytes()
+    else:  # no caching, download every time
+        firmware_data, dl_hash = dl(url)
+
+    return [HexInt(x) for x in firmware_data]
+
+
+def validate_firmware(value):
+    if CONF_URL not in value:
+        try:
+            value[CONF_URL], value[CONF_SHA256] = KNOWN_FIRMWARE[value[CONF_VERSION]]
+        except KeyError as e:
+            raise cv.Invalid(
+                f"Firmware {value[CONF_VERSION]} is unknown, please specify an '{CONF_URL}' ..."
+            ) from e
+    return value
```

Comment:
This should copy the object and return the copy
```suggestion
def validate_firmware(value):
    config = value.copy()
    if CONF_URL not in config:
        try:
            config[CONF_URL], config[CONF_SHA256] = KNOWN_FIRMWARE[config[CONF_VERSION]]
        except KeyError as e:
            raise cv.Invalid(
                f"Firmware {config[CONF_VERSION]} is unknown, please specify an '{CONF_URL}' ..."
            ) from e
    return config
```